cpp/src/io/orc/writer_impl.cu

```diff
@@ -1121,96 +1147,163 @@ writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
     } else {
       desc->ts_scale = 0;
     }
+    col_stats_dtypes.push_back(desc->stats_dtype);
+    col_types.push_back(column.type());
     for (auto const& stripe : segmentation.stripes) {
-      auto& grp = stripe_stat_merge[column.index() * segmentation.num_stripes() + stripe.id];
-      grp.col   = stat_desc.device_ptr(column.index());
+      auto& grp       = stripe_stat_merge[column.index() * segmentation.num_stripes() + stripe.id];
+      grp.col_dtype   = column.type();
+      grp.stats_dtype = desc->stats_dtype;
       grp.start_chunk =
         static_cast<uint32_t>(column.index() * segmentation.num_rowgroups() + stripe.first);
       grp.num_chunks = stripe.size;
       for (auto rg_idx_it = stripe.cbegin(); rg_idx_it < stripe.cend(); ++rg_idx_it) {
         auto& rg_grp =
           rowgroup_stat_merge[column.index() * segmentation.num_rowgroups() + *rg_idx_it];
-        rg_grp.col         = stat_desc.device_ptr(column.index());
+        rg_grp.col_dtype   = column.type();
+        rg_grp.stats_dtype = desc->stats_dtype;
         rg_grp.start_chunk = *rg_idx_it;
         rg_grp.num_chunks  = 1;
       }
     }
-    auto col_stats         = &file_stat_merge[column.index()];
-    col_stats->col         = stat_desc.device_ptr(column.index());
-    col_stats->start_chunk = static_cast<uint32_t>(column.index() * segmentation.num_stripes());
-    col_stats->num_chunks  = static_cast<uint32_t>(segmentation.num_stripes());
   }
   stat_desc.host_to_device(stream);
-  stat_merge.host_to_device(stream);
+  rowgroup_merge.host_to_device(stream);
+  stripe_merge.host_to_device(stream);
   set_stat_desc_leaf_cols(orc_table.d_columns, stat_desc, stream);
 
-  rmm::device_uvector<statistics_chunk> stat_chunks(num_stat_blobs, stream);
-  auto rowgroup_stat_chunks = stat_chunks.data();
-  auto stripe_stat_chunks   = rowgroup_stat_chunks + num_rowgroup_blobs;
-  auto file_stat_chunks     = stripe_stat_chunks + num_stripe_blobs;
+  // rowgroup stat chunks are written out in each stripe stripe and file-level chunks are written in
+  // the footer. to prevent persisting the rowgroup stat chunks across multiple calls to write in a
+  // chunked write situation, these allocations are split up so stripe data can persist until the
+  // footer write
```

Comment:
Tried to improve wording here.
```suggestion
  // The rowgroup stat chunks are written out in each stripe and file-level chunks are written in
  // the footer. To prevent persisting the rowgroup stat chunks across multiple write calls in a
  // chunked write situation, these allocations are split up so stripe data can persist until the
  // footer is written.
```

Reply:
I'm going to commit this and then edit it a bit, would be nice to re-review. The stripe stripe was actually different thoughts. Rowgroup stat chunks are written out in each stripe. And then stripe and file-level chunks are written in the footer. I'll start with yours and attempt to make this more clear.